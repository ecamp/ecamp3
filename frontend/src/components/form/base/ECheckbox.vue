<template>
  <ValidationProvider
    v-slot="{ errors: veeErrors }"
    :name="name"
    :vid="veeId"
    :rules="veeRules"
    :required="required"
  >
    <v-checkbox
      v-bind="$attrs"
      :id="id"
      :hide-details="hideDetails"
      :error-messages="veeErrors.concat(errorMessages)"
      :label="label || name"
      :class="[inputClass]"
      :input-value="value"
      @change="$emit('input', $event)"
      v-on="$listeners"
    >
      <!-- passing through all slots -->
      <slot v-for="(_, name) in $slots" :slot="name" :name="name" />
      <template v-for="(_, name) in $scopedSlots" :slot="name" slot-scope="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </v-checkbox>
  </ValidationProvider>
</template>

<script>
import { ValidationProvider } from 'vee-validate'
import { formComponentPropsMixin } from '@/mixins/formComponentPropsMixin.js'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'

export default {
  name: 'ECheckbox',
  components: { ValidationProvider },
  mixins: [formComponentPropsMixin, formComponentMixin],
  props: {
    value: { type: Boolean, required: false },
  },
}
</script>

<style scoped>
[required] >>> label::after {
  content: '\a0*';
  font-size: 12px;
  color: #d32f2f;
}
[required] >>> .v-input--is-label-active label::after {
  color: gray;
}
</style>
