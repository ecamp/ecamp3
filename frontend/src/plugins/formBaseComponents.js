// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
