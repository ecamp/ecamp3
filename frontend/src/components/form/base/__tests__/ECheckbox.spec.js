import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ECheckbox from '../ECheckbox.vue'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

describe('An ECheckbox', () => {
  let vuetify

  const mount = (options) => {
    const app = Vue.component('App', {
      components: { ECheckbox },
      data: function () {
        return {
          data: null,
        }
      },
      template: `
        <div data-app>
          <e-checkbox v-model="data"/>
        </div>
      `,
    })
    return mountComponent(app, { vuetify, attachTo: document.body, ...options })
  }
  beforeEach(() => {
    vuetify = new Vuetify()
  })
  test('looks like a checkbox', async () => {
    const wrapper = mount()
    await wrapper.setData({ data: false })
    expect(wrapper).toMatchSnapshot('unchecked')

    await wrapper.setData({ data: true })
    expect(wrapper).toMatchSnapshot('checked')
  })

  test('is checked when initialized checked', () => {
    const wrapper = mount({
      data: function () {
        return {
          data: true,
        }
      },
    })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('true')
  })

  test('updates checkbox when vModel changes', async () => {
    const wrapper = mount()
    await wrapper.setData({ data: false })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('false')

    await wrapper.setData({ data: true })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('true')

    await wrapper.setData({ data: false })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('false')
  })

  test('updates vModel when user clicks on checkbox', async () => {
    const wrapper = mount()
    const input = wrapper.find('input')

    await input.trigger('click')
    expect(wrapper.vm.data).toBe(true)

    jest.resetAllMocks()
    await input.trigger('click')
    expect(wrapper.vm.data).toBe(false)
  })
})
