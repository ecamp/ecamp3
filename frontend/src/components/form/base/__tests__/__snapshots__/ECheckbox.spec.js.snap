// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`An ECheckbox > looks like a checkbox > checked 1`] = `
<div data-app>
     <span>
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-input--selection-controls__input">
                              <i aria-hidden="true"
                                 class="v-icon notranslate mdi mdi-checkbox-marked theme--light primary--text"></i>
                              <input aria-checked="true"
                                     id
                                     role="checkbox"
                                     type="checkbox"
                                     value>
                              <div class="v-input--selection-controls__ripple primary--text"></div>
                         </div>
                    </div>
               </div>
          </div>
     </span>
</div>
`;

exports[`An ECheckbox > looks like a checkbox > unchecked 1`] = `
<div data-app>
     <span>
          <div class="v-input v-input--hide-details theme--light v-input--selection-controls v-input--checkbox">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-input--selection-controls__input">
                              <i aria-hidden="true"
                                 class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"></i>
                              <input aria-checked="false"
                                     id
                                     role="checkbox"
                                     type="checkbox"
                                     value>
                              <div class="v-input--selection-controls__ripple"></div>
                         </div>
                    </div>
               </div>
          </div>
     </span>
</div>
`;
